state.go

```diff
@@ -0,0 +1,514 @@
+		return true
+	}
+	return true
+}
+
+func (state *KatzenmintState) Set(key []byte, value []byte) error {
+	state.tree.Set(key, value)
+	return nil
+}
+
+func (state *KatzenmintState) Delete(key []byte) error {
+	_, success := state.tree.Remove(key)
+	if !success {
+		return fmt.Errorf("remove from database failed")
+	}
+	return nil
+}
+
+func (state *KatzenmintState) Get(key []byte) ([]byte, error) {
+	_, val := state.tree.Get(key)
+	if val == nil {
+		return nil, fmt.Errorf("key '%v' does not exist", key)
+	}
+	ret := make([]byte, len(val))
+	copy(ret, val)
+	return ret, nil
+}
+
+// Note: Caller ensures that the epoch is the current epoch +- 1.
+func (state *KatzenmintState) updateMixDescriptor(rawDesc []byte, desc *pki.MixDescriptor, epoch uint64) (err error) {
+	state.Lock()
+	defer state.Unlock()
+
+	pk := desc.IdentityKey.ByteArray()
+
+	// Get the public key -> descriptor map for the epoch.
+	m, ok := state.descriptors[epoch]
+	if !ok {
+		m = make(map[[eddsa.PublicKeySize]byte]*descriptor)
+		state.descriptors[epoch] = m
+	}
+
+	// Check for redundant uploads.
+	if d, ok := m[pk]; ok {
+		if d.raw == nil {
+			return fmt.Errorf("state: Wtf, raw field of descriptor for epoch %v is nil", epoch)
+		}
+		// If the descriptor changes, then it will be rejected to prevent
+		// nodes from reneging on uploads.
+		if !bytes.Equal(d.raw, rawDesc) {
+			return fmt.Errorf("state: Node %v: Conflicting descriptor for epoch %v", desc.IdentityKey, epoch)
+		}
+
+		// Redundant uploads that don't change are harmless.
+		return nil
+	}
+
+	// Ok, this is a new descriptor.
+	if state.documents[epoch] != nil {
+		// If there is a document already, the descriptor is late, and will
+		// never appear in a document, so reject it.
+		return fmt.Errorf("state: Node %v: Late descriptor upload for for epoch %v", desc.IdentityKey, epoch)
+	}
+
+	// Persist the raw descriptor to disk.
+	key := storageKey(descriptorsBucket, desc.IdentityKey.Bytes(), epoch)
+	if err := state.Set(key, rawDesc); err != nil {
+		return err
+	}
+
+	// Store the raw descriptor and the parsed struct.
+	state.deferCommit = append(state.deferCommit, func() {
+		d := new(descriptor)
+		d.desc = desc
+		d.raw = rawDesc
+		m[pk] = d
+	})
+
+	/*
+		id := hex.EncodeToString(desc.IdentityKey.Bytes())
+		fmt.Printf("Node %s: Successfully submitted descriptor for epoch %v.", id, epoch)
+	*/
+	return
+}
+
+// Note: Caller ensures that the epoch is the current epoch +- 1.
+func (state *KatzenmintState) updateDocument(rawDoc []byte, doc *pki.Document, epoch uint64) (err error) {
+	// Cannot lock here
+
+	// Get the public key -> document map for the epoch.
+	m, ok := state.documents[epoch]
+	if ok {
+		if !bytes.Equal(m.raw, rawDoc) {
+			return fmt.Errorf("state: Conflicting document for epoch %v", epoch)
+		}
+		// Redundant uploads that don't change are harmless.
+		return nil
+	}
+
+	e := make([]byte, 8)
+	binary.PutUvarint(e, epoch)
+
+	// Persist the raw descriptor to disk.
```

Comment:
Should the storing-to-disk operation be deferred to commit stage, given that it would not be reversed when a transaction failed?